Drop StringView constructors taking in raw pointers

cdumez opened the pull request:
#### 8763b25f5d245877b35f4729bbea078d9d96b0d5
<pre>
Drop StringView constructors taking in raw pointers
<a href="https://bugs.webkit.org/show_bug.cgi?id=271512">https://bugs.webkit.org/show_bug.cgi?id=271512</a>

Reviewed by Darin Adler and Sihui Liu.

Drop StringView constructors taking in raw pointers, in favor of the ones taking in a std::span.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::notAFunctionSourceAppender):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseDuration):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::checkICULocaleInvariants):
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::canonicalizeTimeZoneName):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::IntlDateTimeFormat::formatToParts const):
(JSC::IntlDateTimeFormat::formatRangeToParts):
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::IntlDurationFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlListFormat.cpp:
(JSC::IntlListFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::IntlLocale::hourCycles):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::formatRangeToPartsInternal):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSRopeString::unsafeView const):
(JSC::JSRopeString::viewWithUnderlyingString const):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser&lt;CharType&gt;::Lexer::lex):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::lexStringSlow):
(JSC::LiteralParser&lt;CharType&gt;::parsePrimitiveValue):
* Source/JavaScriptCore/wasm/WasmIndexOrName.cpp:
(JSC::Wasm::makeString):
* Source/WTF/wtf/HexNumber.cpp:
(WTF::printInternal):
* Source/WTF/wtf/HexNumber.h:
(WTF::HexNumberBuffer::span const):
* Source/WTF/wtf/URL.cpp:
(WTF::URL::setHost):
(WTF::URL::setHostAndPort):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::encodeNonUTF8Query):
(WTF::URLParser::copyURLPartsUntil):
(WTF::URLParser::parsedDataView):
(WTF::URLParser::domainToASCII):
(WTF::URLParser::parseHostAndPort):
* Source/WTF/wtf/cf/CFURLExtras.cpp:
(WTF::isSameOrigin):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::operator StringView const):
(WTF::equal):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::GraphemeClusters::Iterator::Impl::operator* const):
(WTF::convertASCIILowercaseAtom):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::initialize):
(WTF::StringView::StringView):
(WTF::StringView::substring const):
(WTF::StringView::trim const):
* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::setTextForIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIterator::setText):
(WTF::TextBreakIteratorCache::take):
(WTF::CachedTextBreakIterator::CachedTextBreakIterator):
(WTF::CachedLineBreakIteratorFactory::get):
* Source/WTF/wtf/text/cf/TextBreakIteratorCF.h:
(WTF::TextBreakIteratorCF::setText):
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::TextBreakIterator::mapModeToBackingIterator):
(WTF::TextBreakIterator::TextBreakIterator):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::TextBreakIteratorICU):
(WTF::TextBreakIteratorICU::setText):
* Source/WTF/wtf/text/icu/UTextProvider.h:
(WTF::initializeContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::openLatin1UTextProvider):
(WTF::openLatin1ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.h:
* Source/WTF/wtf/text/icu/UTextProviderUTF16.cpp:
(WTF::openUTF16ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderUTF16.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCCertificateGenerator.cpp:
(WebCore::LibWebRTCCertificateGenerator::RTCCertificateGeneratorCallbackWrapper::process):
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
(WebCore::trimInputSample):
(WebCore::WebSocketHandshake::readStatusLine):
(WebCore::WebSocketHandshake::readHTTPHeaders):
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
(WebCore::CSSCustomPropertySyntax::parseComponent):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::mergeIfAdjacent):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::WordAwareIterator::text const):
(WebCore::SearchBuffer::isWordEndMatch const):
(WebCore::SearchBuffer::isWordStartMatch const):
(WebCore::SearchBuffer::search):
* Source/WebCore/editing/TextIterator.h:
(WebCore::TextIteratorCopyableText::text const):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::wordBreakIteratorForMinOffsetBoundary):
(WebCore::wordBreakIteratorForMaxOffsetBoundary):
(WebCore::backwardSearchForBoundaryWithTextIterator):
(WebCore::forwardSearchForBoundaryWithTextIterator):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::append):
* Source/WebCore/html/MediaFragmentURIParser.cpp:
(WebCore::collectFraction):
* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(WebCore::CSSPreloadScanner::emitRule):
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::findBreakIndexSlow):
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::processMeta):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::appendDescriptorAndReset):
(WebCore::parseImageCandidatesFromSrcsetAttribute):
* Source/WebCore/html/track/VTTScanner.cpp:
(WebCore::VTTScanner::scanDigits):
* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::parseParameterName):
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
(WebCore::parseCryptographicDigestImpl):
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::parseEventStreamLine):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseScheme):
(WebCore::ContentSecurityPolicySourceList::parseHost):
(WebCore::ContentSecurityPolicySourceList::parsePath):
(WebCore::ContentSecurityPolicySourceList::parsePort):
* Source/WebCore/platform/Length.cpp:
(WebCore::parseLength):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::offsetForPosition):
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/ComplexTextController.h:
(WebCore::ComplexTextController::ComplexTextRun::span const):
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::ComposedCharacterClusterTextIterator):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::stringWidth):
* Source/WebCore/platform/graphics/TextRun.h:
(WebCore::TextRun::span8):
(WebCore::TextRun::span16):
(WebCore::TextRun::setText):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advance):
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::trimInputSample):
(WebCore::parseHTTPHeader):
* Source/WebCore/platform/network/RFC8941.cpp:
(RFC8941::parseKey):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::constructTextRun):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::capitalize):
(WebCore::RenderText::previousOffset const):
(WebCore::RenderText::previousOffsetForBackwardDeletion const):
(WebCore::RenderText::nextOffset const):
(WebCore::RenderText::stringView const):
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::paintTextAndEmphasisMarksIfNeeded):
* Source/WebCore/svg/SVGLengthList.cpp:
(WebCore::SVGLengthList::parse):
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMInstance::setServerCertificate):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::readSizeFile):
* Source/WebKitLegacy/mac/Misc/WebKitNSStringExtras.mm:
(-[NSString _web_drawAtPoint:font:textColor:]):
(-[NSString _web_widthWithFont:]):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/TextBreakIterator.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/ComplexTextController.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/MonospaceFontTests.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(respondToRangeRequests):

Canonical link: <a href="https://commits.webkit.org/276616@main">https://commits.webkit.org/276616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19a176265fd7fdd66eaed8e0aeed831d4d977782

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47684 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41176 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28464 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21683 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37047 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18145 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18743 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40034 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3215 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38382 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41429 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40346 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49526 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44634 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16678 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21456 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42868 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21811 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51795 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6283 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21139 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10551 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
```diff
@@ -888,7 +888,7 @@ void IntlDateTimeFormat::initializeDateTimeFormat(JSGlobalObject* globalObject,
                 }
                 replaceHourCycleInSkeleton(skeleton, specifiedHour12);
-                dataLogLnIf(IntlDateTimeFormatInternal::verbose, "replaced:(", StringView(skeleton.data(), skeleton.size()), ")");
+                dataLogLnIf(IntlDateTimeFormatInternal::verbose, "replaced:(", StringView(skeleton.span()), ")");
```
Can just pass a `span` directly to `makeString`, no need to use a `StringView`.

cdumez replied:
It didn't work.

darinadler replied:
OK, I guess `dataLog` doesn’t use `StringConcatenate.h` maybe?

darinadler commented on Source/JavaScriptCore/runtime/IntlCollator.cpp:
```diff
@@ -498,7 +498,7 @@ void IntlCollator::checkICULocaleInvariants(const LocaleSet& locales)
                     }
                 } else {
-                    if (StringView(buffer.data(), buffer.size()).containsOnlyASCII()) {
+                    if (StringView(buffer.span()).containsOnlyASCII()) {
```
Can call this directly on a `span`, don’t have to put it into a `StringView`.

darinadler commented on Source/JavaScriptCore/runtime/IntlLocale.cpp:
```diff
@@ -643,7 +643,7 @@ JSArray* IntlLocale::hourCycles(JSGlobalObject* globalObject)
     }
 
-    dataLogLnIf(IntlLocaleInternal::verbose, "pattern:(", StringView(pattern.data(), pattern.size()), ")");
+    dataLogLnIf(IntlLocaleInternal::verbose, "pattern:(", StringView(pattern.span()), ")");
```
Can use `span` directly here, I think. No need for `StringView`.

cdumez replied:
This one didn't work.

darinadler commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -952,7 +952,7 @@ TokenType LiteralParser<CharType>::Lexer::lexStringSlow(LiteralParserToken<CharT
                     for (int i = 1; i < 5; i++) {
                         if (!isASCIIHexDigit(m_ptr[i])) {
-                            m_lexErrorMessage = makeString("\"\\"_s, StringView { m_ptr, 5 }, "\" is not a valid unicode escape"_s);
+                            m_lexErrorMessage = makeString("\"\\"_s, StringView { std::span { m_ptr, 5 } }, "\" is not a valid unicode escape"_s);
```
Can use `span` directly without `StringView`.

darinadler commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -1144,7 +1144,7 @@ ALWAYS_INLINE JSValue LiteralParser<CharType>::parsePrimitiveValue(VM& vm)
         auto tryMakeErrorString = [&] (unsigned length) -> String {
             bool addEllipsis = length != token->stringOrIdentifierLength;
-            return tryMakeString("Unexpected identifier \""_s, StringView { token->identifierStart, length }, addEllipsis ? "..."_s : ""_s, '"');
+            return tryMakeString("Unexpected identifier \""_s, StringView { std::span { token->identifierStart, length } }, addEllipsis ? "..."_s : ""_s, '"');
```
Can use `span` directly without `StringView`.

darinadler commented on Source/JavaScriptCore/wasm/WasmIndexOrName.cpp:
```diff
@@ -59,7 +59,7 @@ String makeString(const IndexOrName& ion)
     if (ion.isIndex())
         return makeString(moduleName, ".wasm-function["_s, ion.index(), ']');
-    return makeString(moduleName, ".wasm-function["_s, StringView(ion.name()->data(), ion.name()->size()), ']');
+    return makeString(moduleName, ".wasm-function["_s, StringView(ion.name()->span()), ']');
```
Can use `span` directly here without `StringView`. I think I fixed this in my patch currently up for review.

darinadler commented on Source/WTF/wtf/text/StringParsingBuffer.h:
```diff
@@ -77,6 +77,17 @@ class StringParsingBuffer final {
+        return { m_position, lengthRemaining() };
+    }
+
+    std::span<const CharacterType> span(size_t count)
```
Is this better than combining `span` and `left`?

szewai commented on Source/WTF/wtf/text/StringParsingBuffer.h:
```diff
@@ -77,6 +77,11 @@ class StringParsingBuffer final {
         return character;
     }
 
+    std::span<const CharacterType> span()
```
Can this be const?

szewai commented on Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
```diff
@@ -170,9 +170,9 @@ String notAFunctionSourceAppender(const String& originalMessage, StringView sour
+        displayValue = StringView(originalMessage.span8().first(notAFunctionIndex - 1));
     else
-        displayValue = StringView(originalMessage.characters16(), notAFunctionIndex - 1);
+        displayValue = StringView(originalMessage.span16().first(notAFunctionIndex - 1));
```
`StringView displayValue = StringView(originalMessage).substring(notAFunctionIndex - 1)?`

cdumez replied:
`.left(notAFunctionIndex - 1)` but yes.

szewai commented on Source/JavaScriptCore/runtime/JSString.h:
```diff
@@ -1100,8 +1100,8 @@ ALWAYS_INLINE StringView JSRopeString::unsafeView(JSGlobalObject* globalObject)
-            return StringView(base.characters8() + substringOffset(), length());
-        return StringView(base.characters16() + substringOffset(), length());
+            return StringView(base.span8().subspan(substringOffset(), length()));
+        return StringView(base.span16().subspan(substringOffset(), length()));
```
`return StringView(base).substring(substringOffset(), length())`?

szewai commented on Source/WTF/wtf/text/StringView.cpp:
```diff
@@ -210,8 +210,8 @@ class StringView::GraphemeClusters::Iterator::Impl {
-            return StringView(m_stringView.characters8() + m_index, m_indexEnd - m_index);
-        return StringView(m_stringView.characters16() + m_index, m_indexEnd - m_index);
+            return StringView(m_stringView.span8().subspan(m_index, m_indexEnd - m_index));
+        return StringView(m_stringView.span16().subspan(m_index, m_indexEnd - m_index));
```
`return m_stringView.substring(m_index, m_indexEnd-m_index)`?

szewai commented on Source/WTF/wtf/text/TextBreakIterator.h:
```diff
@@ -310,7 +310,7 @@ class CachedLineBreakIteratorFactory {
         const UChar* priorContext = m_priorContext.characters();
         if (!m_iterator) {
-            m_iterator = CachedTextBreakIterator(m_stringView, priorContext, m_priorContext.length(), WTF::TextBreakIterator::LineMode { m_mode }, m_locale, m_contentAnalysis);
+            m_iterator = CachedTextBreakIterator(m_stringView, std::span { priorContext, m_priorContext.length() }, WTF::TextBreakIterator::LineMode { m_mode }, m_locale, m_contentAnalysis);
```
Can we just pass `m_priorContext` to `CachedTextBreakIterator`?

cdumez replied:
I doubt it,  `m_priorContext.characters()` is implemented like this:
```
        const UChar* characters() const
        {
            return m_priorContext.data() + (m_priorContext.size() - length());
        }
```

szewai commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -2052,8 +2052,8 @@ StringView RenderText::stringView(unsigned start, std::optional<unsigned> stop)
-        return { text().characters8() + start, destination - start };
-    return { text().characters16() + start, destination - start };
+        return text().span8().subspan(start, destination - start);
+    return text().span16().subspan(start, destination - start);
```
`return StringView(text()).substring(start, destination - start)`?

szewai commented on Source/WebCore/testing/MockCDMFactory.cpp:
```diff
@@ -282,7 +282,8 @@ void MockCDMInstance::initializeWithConfiguration(const MediaKeySystemConfigurat
 {
-    StringView certificateStringView(certificate->makeContiguous()->data(), certificate->size());
+    Ref contiguousData = certificate->makeContiguous();
+    StringView certificateStringView(contiguousData->span());
```
`callback(equalLettersIgnoringASCIICase(contiguousData->span(), "valid"_s) ? Succeeded : Failed);`?

szewai commented on Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
```diff
@@ -186,33 +186,33 @@ TEST(WTF, StringViewIterators)
     EXPECT_TRUE(compareLoopIterations(StringView(b.toString()).codePoints(), {0xD800, 0xD801}));
     EXPECT_TRUE(compareLoopIterations(StringView(b.toString()).codeUnits(), {0xD800, 0xD801}));
-    EXPECT_TRUE(compareLoopIterations(StringView(b.toString()).graphemeClusters(), {StringView(b.characters16(), 1), StringView(b.characters16() + 1, 1)}));
+    EXPECT_TRUE(compareLoopIterations(StringView(b.toString()).graphemeClusters(), { StringView(b.span<UChar>().subspan(0, 1)), StringView(b.span<UChar>().subspan(1, 1)) }));
```
seems we should have `span16()` for consistency

darinadler commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -966,7 +966,7 @@ TokenType LiteralParser<CharType>::Lexer::lexStringSlow(LiteralParserToken<CharT
                         break;
                     }
-                    m_lexErrorMessage = makeString("Invalid escape character "_s, StringView { m_ptr, 1 });
+                    m_lexErrorMessage = makeString("Invalid escape character "_s, StringView { span(*m_ptr) });
```
Don’t need `StringView` here.

darinadler commented on Source/WTF/wtf/text/StringParsingBuffer.h:
```diff
@@ -67,7 +67,7 @@ class StringParsingBuffer final {
     }
 
-    StringView stringViewOfCharactersRemaining() const { return { m_position, lengthRemaining() }; }
+    StringView stringViewOfCharactersRemaining() const { return std::span { m_position, lengthRemaining() }; }
```
Can this just call the new `span` function instead.

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -68,13 +68,12 @@ class StringView final {
-    StringView(const char*, unsigned length);
+    StringView(std::span<const LChar>);
+    StringView(std::span<const UChar>);
+    StringView(std::span<const char>);
```
Might want to remove this eventually since encoding of non-ASCII `char` is ambiguous, unlike `LChar` and `char8_t`.

darinadler commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -656,7 +656,7 @@ static NEVER_INLINE unsigned findBreakIndexSlow(const String& string, unsigned c
     // We need at least two characters look-ahead to account for UTF-16 surrogates.
     unsigned breakSearchLength = std::min(proposedBreakIndex - currentPosition + 2, stringLength - currentPosition);
-    NonSharedCharacterBreakIterator it(StringView(string.characters16(), stringLength).substring(currentPosition, breakSearchLength));
+    NonSharedCharacterBreakIterator it(StringView(string.span16()).substring(currentPosition, breakSearchLength));
```
Make the `StringView` from the `String`? Doesn’t seem like `span16` should be needed.

cdumez replied:
Might have been an optimization to avoid an `is8Bit()` check? Unsure.

darinadler commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -120,7 +120,7 @@ class TokenPreloadScanner::StartTagScanner {
         for (auto& attribute : attributes) {
             auto knownAttributeName = AtomString::lookUp(attribute.name.data(), attribute.name.size());
-            StringView attributeValue { attribute.value.data(), static_cast<unsigned>(attribute.value.size()) };
+            StringView attributeValue { attribute.value.span() };
```
Do we really need the local variable, or any explicit mention of `StringView`?

darinadler commented on Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
```diff
@@ -56,7 +56,7 @@ template<typename CharType>
 {
     if (position > descriptorStart)
-        descriptors.append(StringView(descriptorStart, position - descriptorStart));
+        descriptors.append(StringView(std::span(descriptorStart, position - descriptorStart)));
```
I don’t think we need the explicit `StringView` here. Probably many, **many** other cases like this.

darinadler commented on Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
```diff
@@ -556,7 +556,7 @@ void InbandTextTrackPrivateAVF::processNativeSamples(CFArrayRef nativeSamples, c
                 // A WebVTT header is terminated by "One or more WebVTT line terminators" so append two line feeds to make sure the parser
                 // reccognized this string as a full header.
-                auto header = makeString(StringView { CFDataGetBytePtr(webvttHeaderData), length }, "\n\n");
+                auto header = makeString(StringView { std::span { CFDataGetBytePtr(webvttHeaderData), length } }, "\n\n");
```
No need for `StringView` here.

darinadler commented on Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
```diff
@@ -326,8 +326,8 @@ class ParsedResponseMessage {
-            StringView dataString(reinterpret_cast<const LChar*>(data.data()), data.size());
-            static NeverDestroyed<StringView> type(reinterpret_cast<const LChar*>(":Type:"), 6);
+            StringView dataString(data.span());
+            static NeverDestroyed<StringView> type(":Type:"_s);
```
This seems wasteful. We can just pass `":Type:"_s`.

darinadler commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -318,7 +318,7 @@ static String trimInputSample(CharType* p, size_t length)
     if (length <= maxInputSampleSize)
         return String(p, length);
-    return makeString(StringView(p, length).left(maxInputSampleSize), horizontalEllipsis);
+    return makeString(StringView(std::span { p, length }).left(maxInputSampleSize), horizontalEllipsis);
```
Can just use `first` here and not involve `StringView`.

darinadler reviewed (approved):
Most but not all places we wrote `StringView(std:.span())` can just be `std:span()` since the type is determined by the function we call with the value or other context like assignment or return type.

cdumez committed:
Drop StringView constructors taking in raw pointers
https://bugs.webkit.org/show_bug.cgi?id=271512

Reviewed by Darin Adler and Sihui Liu.

Drop StringView constructors taking in raw pointers, in favor of the ones taking in a std::span.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::notAFunctionSourceAppender):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseDuration):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::checkICULocaleInvariants):
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::canonicalizeTimeZoneName):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::IntlDateTimeFormat::formatToParts const):
(JSC::IntlDateTimeFormat::formatRangeToParts):
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::IntlDurationFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlListFormat.cpp:
(JSC::IntlListFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::IntlLocale::hourCycles):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::formatRangeToPartsInternal):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSRopeString::unsafeView const):
(JSC::JSRopeString::viewWithUnderlyingString const):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser<CharType>::Lexer::lex):
(JSC::LiteralParser<CharType>::Lexer::lexStringSlow):
(JSC::LiteralParser<CharType>::parsePrimitiveValue):
* Source/JavaScriptCore/wasm/WasmIndexOrName.cpp:
(JSC::Wasm::makeString):
* Source/WTF/wtf/HexNumber.cpp:
(WTF::printInternal):
* Source/WTF/wtf/HexNumber.h:
(WTF::HexNumberBuffer::span const):
* Source/WTF/wtf/URL.cpp:
(WTF::URL::setHost):
(WTF::URL::setHostAndPort):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::encodeNonUTF8Query):
(WTF::URLParser::copyURLPartsUntil):
(WTF::URLParser::parsedDataView):
(WTF::URLParser::domainToASCII):
(WTF::URLParser::parseHostAndPort):
* Source/WTF/wtf/cf/CFURLExtras.cpp:
(WTF::isSameOrigin):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::operator StringView const):
(WTF::equal):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::GraphemeClusters::Iterator::Impl::operator* const):
(WTF::convertASCIILowercaseAtom):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::initialize):
(WTF::StringView::StringView):
(WTF::StringView::substring const):
(WTF::StringView::trim const):
* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::setTextForIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIterator::setText):
(WTF::TextBreakIteratorCache::take):
(WTF::CachedTextBreakIterator::CachedTextBreakIterator):
(WTF::CachedLineBreakIteratorFactory::get):
* Source/WTF/wtf/text/cf/TextBreakIteratorCF.h:
(WTF::TextBreakIteratorCF::setText):
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::TextBreakIterator::mapModeToBackingIterator):
(WTF::TextBreakIterator::TextBreakIterator):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::TextBreakIteratorICU):
(WTF::TextBreakIteratorICU::setText):
* Source/WTF/wtf/text/icu/UTextProvider.h:
(WTF::initializeContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::openLatin1UTextProvider):
(WTF::openLatin1ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.h:
* Source/WTF/wtf/text/icu/UTextProviderUTF16.cpp:
(WTF::openUTF16ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderUTF16.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCCertificateGenerator.cpp:
(WebCore::LibWebRTCCertificateGenerator::RTCCertificateGeneratorCallbackWrapper::process):
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
(WebCore::trimInputSample):
(WebCore::WebSocketHandshake::readStatusLine):
(WebCore::WebSocketHandshake::readHTTPHeaders):
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
(WebCore::CSSCustomPropertySyntax::parseComponent):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::mergeIfAdjacent):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::WordAwareIterator::text const):
(WebCore::SearchBuffer::isWordEndMatch const):
(WebCore::SearchBuffer::isWordStartMatch const):
(WebCore::SearchBuffer::search):
* Source/WebCore/editing/TextIterator.h:
(WebCore::TextIteratorCopyableText::text const):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::wordBreakIteratorForMinOffsetBoundary):
(WebCore::wordBreakIteratorForMaxOffsetBoundary):
(WebCore::backwardSearchForBoundaryWithTextIterator):
(WebCore::forwardSearchForBoundaryWithTextIterator):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::append):
* Source/WebCore/html/MediaFragmentURIParser.cpp:
(WebCore::collectFraction):
* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(WebCore::CSSPreloadScanner::emitRule):
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::findBreakIndexSlow):
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::processMeta):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::appendDescriptorAndReset):
(WebCore::parseImageCandidatesFromSrcsetAttribute):
* Source/WebCore/html/track/VTTScanner.cpp:
(WebCore::VTTScanner::scanDigits):
* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::parseParameterName):
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
(WebCore::parseCryptographicDigestImpl):
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::parseEventStreamLine):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseScheme):
(WebCore::ContentSecurityPolicySourceList::parseHost):
(WebCore::ContentSecurityPolicySourceList::parsePath):
(WebCore::ContentSecurityPolicySourceList::parsePort):
* Source/WebCore/platform/Length.cpp:
(WebCore::parseLength):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::offsetForPosition):
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/ComplexTextController.h:
(WebCore::ComplexTextController::ComplexTextRun::span const):
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::ComposedCharacterClusterTextIterator):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::stringWidth):
* Source/WebCore/platform/graphics/TextRun.h:
(WebCore::TextRun::span8):
(WebCore::TextRun::span16):
(WebCore::TextRun::setText):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advance):
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::trimInputSample):
(WebCore::parseHTTPHeader):
* Source/WebCore/platform/network/RFC8941.cpp:
(RFC8941::parseKey):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::constructTextRun):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::capitalize):
(WebCore::RenderText::previousOffset const):
(WebCore::RenderText::previousOffsetForBackwardDeletion const):
(WebCore::RenderText::nextOffset const):
(WebCore::RenderText::stringView const):
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::paintTextAndEmphasisMarksIfNeeded):
* Source/WebCore/svg/SVGLengthList.cpp:
(WebCore::SVGLengthList::parse):
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMInstance::setServerCertificate):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::readSizeFile):
* Source/WebKitLegacy/mac/Misc/WebKitNSStringExtras.mm:
(-[NSString _web_drawAtPoint:font:textColor:]):
(-[NSString _web_widthWithFont:]):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/TextBreakIterator.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/ComplexTextController.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/MonospaceFontTests.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(respondToRangeRequests):

Canonical link: https://commits.webkit.org/276616@main
